src/compiler/checker.ts

```diff
@@ -6258,6 +6258,13 @@ namespace ts {
                     targetType = typeToString(target, /*enclosingDeclaration*/ undefined, TypeFormatFlags.UseFullyQualifiedType);
                 }
 
+                // check if trying to relate primitives to the boxed/apparent backing types.
+                if ((sourceType === "Number" && targetType === "number") ||

```

Comment:
You don't want to do this because you might be in a context where `Number` has been shadowed to something else.

You also need to cover `Symbol`/`symbol`.

Instead, do a check against

* `globalStringType`/`stringType`
* `globalNumberType`/`numberType`
* `globalBooleanType`/`booleanType`
* `getGlobalESSymbolType()`/`esSymbolType`